File: backend/app/api/schemas/harness_checkpoint.py
"""Checkpoint 时间旅行 Pydantic schemas"""
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict


class MessageSnapshotItem(BaseModel):
    """快照中单条消息"""
    model_config = ConfigDict(from_attributes=True)

    id: str
    sender_type: str
    role: str
    content: str = Field(max_length=32000)
    message_type: Optional[str] = "text"
    sent_at: Optional[str] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None
    tool_call_id: Optional[str] = None
    tool_name: Optional[str] = None
    attachments: Optional[List[Dict[str, Any]]] = None
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


class CheckpointResponse(BaseModel):
    """单个 checkpoint 响应"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    conversation_id: UUID
    branch_id: UUID
    parent_checkpoint_id: Optional[UUID]
    step_index: int
    phase: str
    checkpoint_kind: str
    label: Optional[str]
    merge_parents: Optional[List[UUID]]
    is_head: bool
    messages_snapshot: List[MessageSnapshotItem]
    agent_state: Dict[str, Any]
    created_at: str


class BranchResponse(BaseModel):
    """分支响应"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    conversation_id: UUID
    name: str
    parent_branch_id: Optional[UUID]
    head_checkpoint_id: Optional[UUID]
    is_archived: bool
    created_at: str
    closed_at: Optional[str]


# ---- Request ----

class CreateBranchRequest(BaseModel):
    source_checkpoint_id: UUID
    name: str = Field(min_length=1, max_length=100)
    start_with_messages: bool = True


class MergeRequest(BaseModel):
    picked_checkpoint_ids: List[UUID] = Field(min_length=2)
    new_branch_name: str = Field(min_length=1, max_length=100)


class UpdateBranchRequest(BaseModel):
    """PATCH /branches/{id} 请求体（改名 / 归档）"""
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    is_archived: Optional[bool] = None


class WriteCheckpointRequest(BaseModel):
    """POST /checkpoints 请求体（手动写入 checkpoint）"""
    step_index: int
    phase: str
    messages: List[MessageSnapshotItem] = Field(max_length=200)
    scratch_state: Dict[str, Any] = Field(default_factory=dict)
    label: Optional[str] = None


class RollbackResponse(BaseModel):
    conversation_head_checkpoint_id: UUID
    detached_checkpoint_count: int
    target_checkpoint: CheckpointResponse


class CreateBranchResponse(BaseModel):
    """POST /branches 响应：新建分支 + 首个 checkpoint"""
    model_config = ConfigDict(from_attributes=True)

    branch: BranchResponse
    first_checkpoint: CheckpointResponse


class MergeResponse(BaseModel):
    """POST /branches/{id}/merge 响应：新分支 + merge commit"""
    model_config = ConfigDict(from_attributes=True)

    branch: BranchResponse
    merge_commit: CheckpointResponse

File: frontend/src/types/harnessCheckpoint.ts
/**
 * Checkpoint 时间旅行 类型定义
 *
 * Phase 3-Plan-1D / Task 5
 * 字段命名与后端 Pydantic schema 严格对齐（snake_case）。
 */

export interface MessageSnapshotItem {
  id: string;
  sender_type: string;
  role: string;
  content: string;
  message_type?: string | null;
  sent_at?: string | null;
  tool_calls?: unknown[] | null;
  tool_call_id?: string | null;
  tool_name?: string | null;
  attachments?: unknown[] | null;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

export interface Checkpoint {
  id: string;
  conversation_id: string;
  branch_id: string;
  parent_checkpoint_id: string | null;
  step_index: number;
  phase: string;
  checkpoint_kind: 'auto' | 'manual' | 'branch_point' | 'merge_commit';
  label: string | null;
  merge_parents: string[] | null;
  is_head: boolean;
  messages_snapshot: MessageSnapshotItem[];
  agent_state: Record<string, unknown>;
  created_at: string;
}

export interface Branch {
  id: string;
  conversation_id: string;
  name: string;
  parent_branch_id: string | null;
  head_checkpoint_id: string | null;
  is_archived: boolean;
  created_at: string;
  closed_at: string | null;
}

export interface CreateBranchRequest {
  source_checkpoint_id: string;
  name: string;
  start_with_messages: boolean;
}

export interface MergeRequest {
  picked_checkpoint_ids: string[];
  new_branch_name: string;
}

export interface RollbackResponse {
  conversation_head_checkpoint_id: string;
  detached_checkpoint_count: number;
  target_checkpoint: Checkpoint;
}
